driver-core/src/main/com/mongodb/connection/RequestMessage.java

```diff
@@ -141,11 +142,13 @@ public MessageSettings getSettings() {
      * Encoded the message to the given output.
      *
      * @param bsonOutput the output
+     * @param sessionContext the session context
      */
-    public void encode(final BsonOutput bsonOutput) {
+    public void encode(final BsonOutput bsonOutput, final SessionContext sessionContext) {
+        notNull("sessionContext", sessionContext);
```

Comment:
Rubber hits road: the session context is required to be non-null, which forces every level of the stack to pass a non-null session context down.  This is why I think we can get away with not writing unit tests at every level: you can't pass null, so you either have to get it from the binding, the connection source, or explicitly choose to use the NoOpSessionContext.